src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,575 @@
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param algParameterSpec
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        // the lock is not needed, because the Spi will already be set in
+        // chooseProvider
+        lock = null;
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        firstService = s;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * <p>This is the same name that was specified in one of the
+     * {@code getInstance} calls that created this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        chooseFirstProvider();
+        return provider.getName();
+    }
+
+    /**
+     * Creates an instance of the {@code KDF} object.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm
+     * @throws NullPointerException
+     *     if the algorithm is {@code null}
+     */
+    public static KDF getInstance(String algorithm)
+        throws NoSuchAlgorithmException {
+        try {
+            return getInstance(algorithm, (AlgorithmParameterSpec) null);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "Received an InvalidParameterSpecException. Does this "
+                + "algorithm require an "
+                + "AlgorithmParameterSpec?");
+        }
+    }
+
+    /**
+     * Creates an instance of the {@code KDF} object with a specific
+     * {@code Provider}.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param provider
+     *     the provider to use for this key derivation
```

Comment:
It looks like provider can be null. If so, you should specify that, e.g. "If null, this method is equivalent to getInstance(String)."